dbcrossbarlib/src/drivers/gs/mod.rs

```diff
@@ -77,165 +80,21 @@ impl Locator for GsLocator {
-                    &file_url[url.as_str().len()..]
-                } else {
-                    return Err(format_err!(
-                        "expected {} to start with {}",
-                        file_url,
-                        url
-                    ));
-                };
-
-                // Now strip any extension.
-                let name = basename_or_relative
-                    .splitn(2, '.')
-                    .next()
-                    .ok_or_else(|| format_err!("can't get basename of {}", file_url))?
-                    .to_owned();
-                let ctx =
-                    ctx.child(o!("stream" => name.clone(), "url" => file_url.clone()));
-                debug!(ctx.log(), "streaming from `gsutil cp`");
-
-                // Stream the file from the cloud.
-                let mut child = Command::new("gsutil")
-                    .args(&["cp", file_url.as_str(), "-"])
-                    .stdout(Stdio::piped())
-                    .spawn_async()
-                    .context("error running gsutil")?;
-                let child_stdout =
-                    child.stdout().take().expect("child should have stdout");
-                let data = copy_reader_to_stream(ctx.clone(), child_stdout)?;
-                ctx.spawn_process(format!("gsutil cp {} -", file_url), child);
-
-                // Assemble everything into a CSV stream.
-                Ok(CsvStream {
-                    name,
-                    data: Box::new(data),
-                })
-            },
-        )
-        .into_boxed()
-    });
-
-    Ok(Some(Box::new(csv_streams) as BoxStream<CsvStream>))
-}
 
-async fn write_local_data_helper(
-    ctx: Context,
-    url: Url,
-    _schema: Table,
-    data: BoxStream<CsvStream>,
-    if_exists: IfExists,
-) -> Result<BoxStream<BoxFuture<()>>> {
-    // Delete the existing output, if it exists.
-    if if_exists == IfExists::Overwrite {
-        // Delete all the files under `self.url`, but be careful not to
-        // delete the entire bucket. See `gsutil rm --help` for details.
-        debug!(ctx.log(), "deleting existing {}", url);
-        assert!(url.path().ends_with('/'));
-        let delete_url = url.join("**")?;
-        let status = Command::new("gsutil")
-            .args(&["rm", "-f", delete_url.as_str()])
-            .status_async()
-            .context("error running gsutil")?;
-        if !await!(status)?.success() {
-            warn!(
-                ctx.log(),
-                "can't delete contents of {}, possibly because it doesn't exist", url
-            );
-        }
-    } else {
-        return Err(format_err!(
-            "must specify `overwrite` for gs:// destination"
-        ));
+    fn supports_write_remote_data(&self, source: &dyn Locator) -> bool {
+        // We can only do `write_remote_data` if `source` is a `BigQueryLocator`.
+        // Otherwise, we need to do `write_local_data` like normal.
+        source.as_any().is::<BigQueryLocator>()
```

Comment:
The `write_remote_data` stuff is for moving directly between two cloud locations (BigQuery and Google Cloud Storage) that know how to talk to each other without us having to download the CSVs to the local machine at all.

This is commented on the abstract `Locator` trait, and the Rust doc tool will automatically copy that documentation to each of the local implementations, like this one here, when we render the API docs. So mostly I don't document the local implementations.